Data Union fixes ported from v5

jtakalai opened the pull request.

jtakalai committed:
docs: cherry-picked TODO comments from ETH-216 branch

jtakalai committed:
fix: query the sidechain address directly from mainnet contract, NOT from factory

See commit 4717adf13c769048aacee036779fea4bfc64ecd6 where this change was made to v5

jtakalai committed:
test: enable DU tests

jtakalai committed:
fix: getStats

ported from v5

jtakalai committed:
feat: safeGetDataUnion → getDataUnion

this kind of became necessary for this branch so cherry-picked it

jtakalai committed:
test: more getDataUnion fixes

this belongs to ETH-217 really

jtakalai committed:
refactor: removed  _fromContractAddress and _fromName

those used the calculated stuff; better just inline them into the few places they were needed

jtakalai committed:
test: fixes

harbu committed:
Revert "test: fixes"

This reverts commit 716e1e2b63c259b6bc2f2bb184d7cdcc97d715a2.

harbu committed:
Merge branch 'main' into ETH-218-port-v5-fixes-to-v6

harbu committed:
fix eslint

harbu committed:
Revert "Revert "test: fixes""

This reverts commit 9c554cce86e6074252f5a1de9fce6e19f30ef4ce.

harbu committed:
Revert "test: enable DU tests"

This reverts commit 1685018e632052ad21c81cf710443e86c99fbc08.

jtakalai committed:
Merge remote-tracking branch 'origin/main' into ETH-218-port-v5-fixes-to-v6

# Conflicts:
#	packages/client/src/dataunion/Contracts.ts
#	packages/client/src/dataunion/index.ts

jtakalai committed:
Merge branch 'ETH-218-port-v5-fixes-to-v6' of github.com:streamr-dev/network-monorepo into ETH-218-port-v5-fixes-to-v6

jtakalai committed:
fix: admin fee is now in sidechain contract

jtakalai committed:
test: getAddress now returns all lowercase

jtakalai committed:
test: getDataUnion is now async

harbu commented on packages/client/src/dataunion/Contracts.ts:
```diff
@@ -56,60 +56,82 @@ export default class Contracts {
-     * TODO: rename this to calculateDataUnionMainnetAddress, see v5 branch, around PR #285
      */
-    getDataUnionMainnetAddress(dataUnionName: string, deployerAddress: EthereumAddress) {
+    calculateDataUnionMainnetAddress(dataUnionName: string, deployerAddress: EthereumAddress, templateMainnetAddress?: EthereumAddress) {
```
It can be useful to be explicit about the return type when declaring a method / function, especially if it is a public one. It serves as documentation to a reader (I know the return type gets inferred but sometimes it can be unruly what gets generated) and it also prevents a certain type of errors I have seen where the author has ended up with a return type they probably didn't intend.

This is a bit artificial, but something similar in nature that I encountered a while ago:
```ts
function positiveAdd(a: number, b:number) {
        if (a < 0 || b < 0) { return }
        return a+ b
    }
 ```
 Quickly glancing, you may expect the return type to be `number` but it is `number | undefined`. By being explicit about the return type
 ```ts
 function positiveAdd(a: number, b: number): number {
    if (a < 0 || b < 0) { return }
    return a + b
}
 ```
 you get compilation error 
 ```
 TS2322: Type 'undefined' is not assignable to type 'number'.
 ```

harbu commented on packages/client/test/integration/dataunion/calculate.test.ts:
```diff
@@ -24,14 +26,17 @@ describe('DataUnion calculate', () => {
-        // eslint-disable-next-line no-underscore-dangle
-        const dataUnionPredicted = adminClient._getDataUnionFromName({ dataUnionName, deployerAddress: adminWalletMainnet.address })
+
+        // @ts-expect-error
```
A good idea is to add a short description after `@ts-expect-error` to indicate to the future reader why it exists in the first place.

E.g.
```
// @ts-expect-error accessing private method
```

harbu reviewed (approved):
LGTM; merge at your convenience when tests are green

jtakalai committed:
Merge branch 'main' into ETH-218-port-v5-fixes-to-v6

jtakalai committed:
build: run jest from vscode

use jest in node_modules instead of global installation

jtakalai committed:
fix: ABI update

jtakalai committed:
refactor: eliminate internal _getContracts

inline into tests that need it

jtakalai committed:
test: add session to fetch

apparently it needs it now

jtakalai committed:
test: member join

I think the tests assume the member has joined too and will receive tokens

jtakalai committed:
test: lint fix

but also probably test fix because full ABI is needed here probably

jtakalai committed:
Merge branch 'main' into ETH-218-port-v5-fixes-to-v6

jtakalai committed:
refactor: add types

PR review feedback

jtakalai committed:
test: fix BigNumber incompatibility

beats me where that can come from, but apparently there were two kinds of BigNumbers and I got errors like
```
    expect(received).toMatchObject(expected)

    - Expected  - 4
    + Received  + 4

    @@ -3,13 +3,13 @@
          "hex": "0x00",
          "type": "BigNumber",
        },
        "status": "ACTIVE",
        "totalEarnings": Object {
    -     "hex": "0x0de0b6b3a7640000",
    -     "type": "BigNumber",
    +     "_hex": "0x06f05b59d3b20000",
    +     "_isBigNumber": true,
        },
        "withdrawableEarnings": Object {
    -     "hex": "0x0de0b6b3a7640000",
    -     "type": "BigNumber",
    +     "_hex": "0x06f05b59d3b20000",
    +     "_isBigNumber": true,
        },
      }
```

jtakalai committed:
test: fix signature test

expected signature changed because sidechain DU address used to be whatever 0x22... mainnet address produces; now sidechain DU address is simply 0x22...

jtakalai committed:
test: fix types

also resolves a PR feedback

jtakalai committed:
test: fix types

avoid unnecessary ts-expect-error

jtakalai committed:
build: fix launch.json

revert back the useful parts

jtakalai committed:
test: fix test
